packages/dynamic-import/server.js

```diff
@@ -1,48 +1,129 @@
-import { readFileSync } from "fs";
-import {
-  join as pathJoin,
-  normalize as pathNormalize,
-} from "path";
+"use strict";
 
-import { check } from "meteor/check";
+const assert = require("assert");
+const { readFileSync } = require("fs");
+const {
+  join: pathJoin,
+  normalize: pathNormalize,
+} = require("path");
+const hasOwn = Object.prototype.hasOwnProperty;
 
-import "./security.js";
-import "./client.js";
+require("./security.js");
 
-const hasOwn = Object.prototype.hasOwnProperty;
+const client = require("./client.js");
+
+Meteor.startup(() => {
+  if (! Package.webapp) {
+    // If the webapp package is not in use, there's no way for the
+    // dynamic-import package to fetch dynamic modules, so we should
+    // abandon the rest of the logic in this module.
+    //
+    // If api.use("webapp") appeared in dynamic-import/package.js, then
+    // Package.webapp would always be defined here, of course, but that
+    // would be a bad idea, because the dynamic-import package should not
+    // single-handedly force a dependency on webapp if the program does
+    // not otherwise need a web server (e.g., when the program is an
+    // isopacket or build plugin instead of a web application).
+    //
+    // Note that the client.js module (imported above) still defines
+    // Module.prototype.dynamicImport, which will work as long as no
+    // modules need to be fetched.
+    return;
+  }
 
-Object.keys(dynamicImportInfo).forEach(platform => {
+  Object.keys(dynamicImportInfo).forEach(setUpPlatform);
+
+  Package.webapp.WebApp.connectHandlers.use(
+    "/__dynamicImport",
+    middleware
+  );
+});
+
+function setUpPlatform(platform) {
   const info = dynamicImportInfo[platform];
+
   if (info.dynamicRoot) {
     info.dynamicRoot = pathNormalize(info.dynamicRoot);
   }
-});
 
-Meteor.methods({
-  __dynamicImport(tree) {
-    check(tree, Object);
-    this.unblock();
-
-    const platform = this.connection ? "web.browser" : "server";
-    const pathParts = [];
-
-    function walk(node) {
-      if (node && typeof node === "object") {
-        Object.keys(node).forEach(name => {
-          pathParts.push(name);
-          node[name] = walk(node[name]);
-          assert.strictEqual(pathParts.pop(), name);
-        });
-      } else {
-        return read(pathParts, platform);
+  if (platform === "server") {
+    client.setSecretKey(info.key = randomId(40));
+  }
+}
+
+function randomId(n) {
+  let s = "";
+  while (s.length < n) {
+    s += Math.random().toString(36).slice(2);
+  }
+  return s.slice(0, n);
+}
+
+function middleware(request, response) {
+  assert.strictEqual(request.method, "POST");
+  const chunks = [];
+  request.on("data", chunk => chunks.push(chunk));
+  request.on("end", () => {
+    response.setHeader("Content-Type", "application/json");
+    response.end(JSON.stringify(readTree(
+      JSON.parse(Buffer.concat(chunks)),
+      getPlatform(request)
+    )));
+  });
+}
+
+function getPlatform(request) {
```

Comment:
Some of the changes in this file revolve around the fact that a configurable `platform` is now supported. Should we add an extra comment somewhere explaining why? Maybe explaining this is better suited elsewhere (like the package README or API docs), but I just thought I'd mention it.

Reply:
We've always had `web.browser` and `server` as separate platforms, it's just that we were able to determine which one to use based on `this.connection` in the method implementation before, and now we have to detect the platform differently. The detection is still fully automatic, so I don't know that it needs to be advertised as a public API. More implementation comments could be nice, though!

Side note: I suppose `web.cordova` could also appear here, though we intentionally force dynamic modules into the main JS bundle for Cordova, because we don't have to keep re-downloading the bundle (so it's ok if it's bigger), and we can't necessarily `eval` dynamic code at runtime in some Cordova environments, due to security restrictions.

Reply:
Make sense @benjamn - looks good as it is.